### PR TITLE
BioCoder integration

### PR DESCRIPTION
**Pull request by tangxiangru**
Integrates BioCoder from https://arxiv.org/abs/2308.16458

### CODE REVIEW

**Comment by lilbillybiscuit** [comment]
### Description:

This pull request is to introduce an evaluation on the BioCoder dataset using agents on OpenDevin. The Biocoder benchmark aims to assess the performance of LLMs on entire repositories related to bioinformatics. It is a challenging full-repository level benchmark that requires both understanding of long-range, inter-class dependencies and domain knowledge.

By evaluating this dataset, we can gain further insights on the performance of LLMs on closed domain (bioinformatics) tasks, as well as their ability to retrieve information and relevant code from a large context window, or the entire repository.

### Key changes:
Added `evaluation/biocoder/run_infer.py`: This file handles execution of the entire Biocoder benchmark using OpenDevin. It loads datasets, runs the agent on each task, and saves the evaluation outputs. Note that we are still 
Added `evaluation/biocoder/biocoder_env_box.py`: This file handles setting up the Biocoder docker environment (which contains many dependencies, libraries, and environments shared across the repositories in Biocoder, including downloading repository files and metadata.

Updated `README.md`: We have updated the README.md to include information about Biocoder, such as the design, challenges, and objectives of the benchmark. Furthermore, we include documentations on executing the framework, an example of the evaluation, and instructions for contributing and customization.

### Related papers:
BioCoder: A Benchmark for Bioinformatics Code Generation with Large Language Models (https://arxiv.org/pdf/2308.16458)

File: evaluation/swe_bench/swe_env_box.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def __init__(
             container_image is not None
         ), 'container_image is required for SWEBenchSSHBox!'
         # Need to run as root to use SWEBench container
-        sid = f'swe_bench_{swe_instance_id}' + str(uuid.uuid4())
+        sid = f'swe_bench_{swe_instance_id}' + str(uuid.uuid4())  # unique session id
```
Comment (yufansong) [inline evaluation/swe_bench/swe_env_box.py:39]:
```suggestion
        sid = f'swe_bench_{swe_instance_id}' + str(uuid.uuid4())
```

File: evaluation/biocoder/README.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# HumanEvalFix Evaluation with BioCoder
```
Comment (yufansong) [inline evaluation/biocoder/README.md:1]:
```suggestion
# BioCoder Evaluation with Opendevin
```

File: evaluation/biocoder/README.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+## Setup Environment
+
+Please follow [this document](https://github.com/OpenDevin/OpenDevin/blob/main/Development.md) to setup local develop environment for OpenDevin.
+
+
+## Configure OpenDevin and your LLM
+
+
+## Run Inference 
+
```
Comment (yufansong) [inline evaluation/biocoder/README.md:14]: Can you add more details about how to run your benchamrk and also add `run_infer.sh`?

**Review by yufansong — COMMENTED** [review]
I am little confuse What is `BiocoderSSHBox` for? It seems not called in `run_infer.py`.

**Comment by neubig** [comment]
@tangxiangru and @lilbillybiscuit , thanks a lot for contributing this! Could you ping us again when the PR is finished and ready for review (including the documentation of how to run)? I'll change it to a draft PR in the meantime.

File: evaluation/biocoder/run_infer.py [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+            except ValueError:
+                # skip this line
+                print(f'Error parsing result line: {line}')
+                continue
+            value = value.strip()
+            try:
+                value = int(value)
+            except ValueError:
+                pass
+            test_result['result'][key.strip()] = value
```
Comment (li-boxuan) [inline evaluation/biocoder/run_infer.py:183]: Please also add metrics, see https://github.com/OpenDevin/OpenDevin/pull/2199 for example

**Review by neubig — COMMENTED** [review]
Hi @tangxiangru , @lilbillybiscuit, and @li-boxuan -- now that we're done with the main evals for the paper it'd be great to get this merged if you have a moment!

**Review by yufansong — CHANGES_REQUESTED** [review]
@tangxiangru  @lilbillybiscuit I try to help you fix some comments and add shell script and readme file. But when I try to run your code locally, I actually meet some problem around the `instance` and `BiocoderData`. When I fix one, there are some follow up problems. So I guess you have not push the final version evalution code here.

So please ping me to review when you finish the code. Thanks.

**Comment by lilbillybiscuit** [comment]
Hi @yufansong and @tangxiangru , I am sorry for the delay, but the final version has been pushed. I am writing a readme but the script can be run with the same arguments as SWE-bench. I am currently writing a more polished readme and it will be pushed shortly.

To answer the previous question, BiocoderSSHBox is a modified version of SWEBenchSSHBox that runs our docker image and all the dependencies (such as fetching the repository archive, testing caches, etc.)

**Review by yufansong — APPROVED** [review]

**Pull request merged**